fix: create `working_dir` for `web` if it doesn't exist, fixes #6158

stasadev opened the pull request:
## The Issue

- https://github.com/phenaproxima/starshot-prototype/issues/94
- #6158

If there is a config in `.ddev/config.yaml`:

```yaml
working_dir:
  web: /var/www/html/something
```

And the `something` directory doesn't exist on the host side, it will be created by Docker as root.

## How This PR Solves The Issue

We can create the directory beforehand and it is only needed if it is in the approot.

## Manual Testing Instructions

1. Use Linux/WSL2.
2. `ddev config global --performance-mode=none`
3. `ddev config --web-working-dir=/var/www/html/something`
4. `ddev start`
5. check file permissions `ls -l something`

## Automated Testing Overview

<!-- Please describe the tests introduced by this PR, or explain why no tests are needed. -->

## Related Issue Link(s)

- https://github.com/moby/moby/issues/2259

## Release/Deployment Notes

<!-- Does this affect anything else or have ramifications for other code? Does anything have to be done on deployment? -->

rfay commented:
I guess our quickstarts used to say `--create-docroot` and now they do it without even asking for that. So we always create the docroot in advance. 

Adding other mounts with a `docker-compose.mounts.yaml` would be an interesting test. 

Also, we normally mount the upload_dirs  when using mutagen. We should be having trouble with that in this situation. But of course there we already have created the host-side directory also (because it can't be mounted otherwise)

rfay commented:
Rebased.

rfay reviewed (changes requested):
I would prefer that we do this work in a separate function. Even though GetWorkingDir() isn't called a lot of places, the only place we need to do this work is in a function called right before RenderComposeYaml(). 

Thanks for your flexibility :)

We should probably have long ago put "/var/www/html" into a constant so it could be used more reliably. We don't have to do that here.

rfay commented:
There's another problem here. If you `ddev config --web-working-dir=/var/www/html/some/deep/dir` then `some` and `some/deep` and `some/deep/dir` all get root ownership on Linux. Maybe the existing approach here is fine and solves that though.

stasadev commented:
> Maybe the existing approach here is fine and solves that though.

Yes, this works for nested directories as well.

rfay commented on pkg/ddevapp/utils.go:
```diff
@@ -522,6 +522,27 @@ func (app *DdevApp) CanUseHTTPOnly() bool {
+// This problem (particularly for Docker volumes) is described in
+// https://github.com/moby/moby/issues/2259
+func (app *DdevApp) CreateWebWorkingDir() {
+	if app.WorkingDir != nil {
```
Do we need to do any of this logic, or just use the working dir whatever it is, and os.MkdirAll() whatever it is? Since it mostly exists, it will do no harm right?

rfay replied:
BTW, this organization is what I was hoping for, thanks.

stasadev replied:
Yes, we need to do all this logic because it's done *outside* the container, before `docker-compose build`.

And yes, we need to check that `working_dir` is set to some path inside `/var/www/html/`, otherwise we won't have a corresponding directory on the host side.

rfay replied:
I would be more comfortable with this if it were encapsulated a bit more. How about creating an app.GetHostWorkingDir(servicename) that does what needs to be done, removing the proper prefix. Then this code here would just be
```
d := GetHostWorkingDir("web")
os.MkDirAll(d, 0755)
```

I think it would be easier to follow.  

Another option is to return two values (in-container and on-host) from app.GetWorkingDir(). It's only used 5 places, so might not be too painful.

stasadev commented:
I refactored it to `app.GetHostWorkingDir`, and even found a bug in my previous code: some project types had a specific `docrootWorkingDir` that overrides the default, which leads to a different prefix, so I switched to using `app.GetAbsAppRoot`.

gitressa commented:
This is great, thanks! I look forward to trying it, as soon as the DDEV install for Drupal Starshot gets fixed.

- https://github.com/phenaproxima/starshot-prototype/issues/100
